test/src/test/java/org/corfudb/runtime/CorfuRuntimeTest.java

```diff
@@ -169,8 +195,70 @@ public void doesNotAllowReadsAfterSealAndBeforeNewLayout() throws Exception {
 
         LogUnitClient luc = runtime.getRouter(SERVERS.ENDPOINT_0).getClient(LogUnitClient.class);
 
-        assertThatThrownBy(() ->luc.read(0).get())
+        assertThatThrownBy(() -> luc.read(0).get())
                 .isInstanceOf(ExecutionException.class)
                 .hasRootCauseInstanceOf(WrongEpochException.class);
     }
+
+    /**
+     * Creates and bootstraps 3 nodes N0, N1 and N2.
+     * The runtime connects to the 3 nodes and sets the clientRouter epochs to 1, 1 and 1.
+     * Now the epoch is updated to 2.
+     * We now add a rule which throws a NetworkException while fetching a router to the first node
+     * in the list. The runtime is now invalidated, which forces to update the client router
+     * epochs. This should now update the epochs to the following: 1, 2, 2.
+     *
+     * @throws Exception
+     */
+    @Test
+    public void testFailedRouterInFetchLayout() throws Exception {
+
+        final Map<String, TestClientRouter> routerMap = new ConcurrentHashMap<>();
+        final AtomicReference<String> failedNode = new AtomicReference<>();
+
+        CorfuRuntime.overrideGetRouterFunction = (corfuRuntime, endpoint) -> {
+            if (failedNode.get() != null && endpoint.equals(failedNode.get())) {
```

Comment:
Can you just move this logic to AbstractViewTest::getRouterFunction?

Reply:
Ah, I see what is happening here. Is there a reason to throw these exceptions (it seems like they aren't picked up in your test?)

Reply:
Never mind, I see the logic. It's an unfortunate side effect of the test router not being able to simulate a closed connection. Add that to the comments (that this is simulating a failed node), and I think we'll be okay.

Reply:
Firstly the test routers never throw Network Exceptions and I needed to test them specifically.
These exceptions are caught by the CorfuRuntime while updating the router epochs.

This is very specific to this test and I wonder if anyone else would ever want to throw this exception at this point in the code.

Reply:
I introduced a "simulated network exception" at the test router level. Can you not leverage this? It is way less fancy than that, so not sure it serves your purpose. That is what this method does, throwing a network exception from the router.

The method is simulateDisconnectedEndpoint(), in TestClientRouter.

Reply:
Sounds good I'll add a dependency on #982 and consume your test addition.